Console v1.3

Atomfusion1 opened the pull request:
This Update is for the Console window, Adding additional functionality

Added Settings under Windows-> Console
A. Lock Window
B. Hide TitleBar
C. Window Alpha Normal
D. Window Alpha Fade
E. Console Font Size Adjustment
F. UI Selector under console (Can Select Fonts for all Windows and Console text
Files Changed ImGuiUtils:

Added 2 Additional Font Sizes of Lucon 15 and 17
MQ2ImGuiConsole:
GetConsolesFlags() Function to handle Flags
MouseOverUpdate() Function to handle Window Fade Delay from Normal to Fade Condition
Set m_zepEditor to use Global Font with Selectable Size
FontNumberFromString() Get Front[#] from Debug String file name
Created ImGui Items for Console Window, 2 check boxes 3 sliders 1 font control
Save All Changes to MacroQuest.ini
Changed Save Location from [Macroquest] to [MQConsole] As now there are 7+ options
Saving: ConsoleFontOnStartup, ShowMacroQuestConsole, LockConsoleWindow, NoTitlebar, ConsoleFontSize, AlphaFadeOnStartup, AlphaNormalOnStartup
Autoload all options on Startup
Set Default Values for All Options if Button Pressed

All previous Pull Request have been Changed/Attempted to be fixed or modified. 
There is still one Issue That i am having with setup and use of accessor in MQ2ImGuiConsole.cpp line 879

brainiac commented on src/main/MQ2ImGuiConsole.cpp:
```diff
@@ -569,6 +579,21 @@ class ZepMode_ImGuiConsole : public Zep::ZepMode_Standard
+	else
+	{
+		auto m_mouseOverDuration = duration_cast<seconds>(steady_clock::now() - s_mouseHoverStartTimePoint);
+		return (m_mouseOverDuration < s_mouseHoverTimerSecondsDelay) ? true : false;
```
using ternary for boolean expression is completely unnecessary. The condition itself is a boolean, so just return it.

brainiac commented on src/main/MQ2ImGuiConsole.cpp:
```diff
@@ -569,6 +579,21 @@ class ZepMode_ImGuiConsole : public Zep::ZepMode_Standard
 	}
 };
 
+bool TimedMouseHoveredEvent(bool m_isMouseHovered)
```
`m_` is the prefix for a member variable, of which this is not.

brainiac commented on src/main/MQ2ImGuiConsole.cpp:
```diff
@@ -64,6 +64,16 @@ static bool s_consoleVisible = false;
+static int s_consoleFontSize = 13;
+static auto s_mouseHoverTimerSecondsDelay = std::chrono::seconds(10);
+static std::chrono::steady_clock::time_point s_mouseHoverStartTimePoint;
+static std::string s_consoleFontName;
```
I suspect a lot of these should be member variables. The reason why some of these exist as statics is because they hold state external to the console that is used to control it.

Most of this seems internal to the console and don't need to be here.

brainiac commented on src/imgui/ImGuiUtils.cpp:
```diff
@@ -86,6 +86,8 @@ void ConfigureDefaultFont(ImFontAtlas* atlas)
 			consoleFontConfig.OversampleH = consoleFontConfig.OversampleV = 3;
 			ConsoleFont = atlas->AddFontFromFileTTF(szConsoleFont, 13.0f, &consoleFontConfig);
+			atlas->AddFontFromFileTTF(szConsoleFont, 15.0f, &consoleFontConfig);
+			atlas->AddFontFromFileTTF(szConsoleFont, 17.0f, &consoleFontConfig);
```
creating these and never using them is a significant waste.

brainiac commented on src/main/MQ2ImGuiConsole.cpp:
```diff
@@ -947,21 +974,31 @@ class ImGuiConsole
+		if (s_isTitleBarShown)  windowFlags |= ImGuiWindowFlags_NoTitleBar;
+		if (s_isConsoleLocked) windowFlags |= ImGuiWindowFlags_NoMove | ImGuiWindowFlags_NoResize;
+		return windowFlags;
+	}
```
this still doesn't warrant being a function.

brainiac commented on src/main/MQ2ImGuiConsole.cpp:
```diff
@@ -1054,6 +1092,12 @@ class ImGuiConsole
+		// Set Font for ZepEditor
+		if (s_updateConsoleFont)
+		{
+			m_zepEditor->SetFont(Zep::ZepTextType::Text, ImGui::GetFont(), s_consoleFontSize);
```
the console specifically uses the console font. This gets the default font. Theres a big mismatch here in the logic.

brainiac commented on src/main/MQ2ImGuiConsole.cpp:
```diff
@@ -947,21 +974,31 @@ class ImGuiConsole
 		ImGui::SetNextWindowSize(ImVec2(640, 240), ImGuiCond_FirstUseEver);
 		ImGui::PushStyleVar(ImGuiStyleVar_WindowPadding, ImVec2(1, 0));
-
+		ImGui::SetNextWindowBgAlpha(.01f * (s_isMouseInFadeDelay ? s_alphaConsoleNormal : s_alphaConsoleFade));
```
if this doesn't need fading we probably can skip calling this at all

brainiac commented on src/main/MQ2ImGuiConsole.cpp:
```diff
@@ -1068,13 +1112,11 @@ class ImGuiConsole
 		ImGui::SetCursorPosY(ImGui::GetCursorPosY() + 4);
 		ImGui::PushItemWidth(ImGui::GetContentRegionAvailWidth());
 		ImGui::PushStyleColor(ImGuiCol_FrameBg, ImVec4(0.00f, 0.00f, 0.00f, 0.00f));
-		ImGui::PushFont(mq::imgui::ConsoleFont);
```
As mentioned above, you've got a logic error in your understanding of how the fonts are being used. We use a variable-width font for all the UI and a monospace font for the console. With your changes, everything is now using the monospace font which is not ok.

If you want to change fonts, you need to change the console font.

brainiac reviewed (changes requested):
There are a LOT of font related issues. I suggest taking out the font changes and focusing on one thing at a time.

brainiac commented:
For the hover thing that we talked about briefly:

The check for being hovered can go where it needs in Render, but it should assign its value to a member variable of ImGuiZepEditor, say `bool m_hovered`

Then, in the class for ImGuiZepEditor add an accessor to retrieve its value

`bool IsHovered() const { return m_hovered; }`

Atomfusion1 commented on src/main/MQ2ImGuiConsole.cpp:
```diff
@@ -64,6 +64,15 @@ static bool s_consoleVisible = false;
 static bool s_setFocus = false;
+static bool s_isConsoleLocked = false;
+
+static bool s_isTitleBarShown = false;
```
My Problem Is I cant not figure out how to Move these into Member for instance Into Line 938, ImGuiConsole 
int m_alphaConsoleNormal;
And access it in 1544 	if (ImGui::SliderInt("##Alpha Normal", &m_alphaConsoleNormal, 0, 100))
	{
		WritePrivateProfileInt("MQConsole", "AlphaNormalOnStartup", m_alphaConsoleNormal, mq::internal_paths::MQini);
	}

Atomfusion1 commented on src/main/MQ2ImGuiConsole.cpp:
```diff
@@ -1055,7 +1087,11 @@ class ImGuiConsole
+		imgui::ImGuiZepEditor obj;
+//		s_isMouseInFadeDelay = TimedMouseHoveredEvent(obj.IsHovered());
+		s_isMouseInFadeDelay = TimedMouseHoveredEvent(ImGui::IsWindowHovered());
+		ImGui::EndChild();
```
I Keep trying but this Works This way but does not work With the comment line above it, I know this is a scope issue, but i have not been able to access it either Directly or through ImGuiConsole::IsHovered() .. This seems to always return true

Atomfusion1 commented:
Still have Git Wrong

Atomfusion1 commented:
This Update is for the Console window, Adding additional functionality

Added Settings under Windows-> Console
A. Lock Window
B. Hide TitleBar
C. Window Alpha Normal
D. Window Alpha Fade

MQ2ImGuiConsole:
TimedMouseHoveredFader Function to handle Window Fade Delay from Normal to Fade Condition
Created ImGui Items for Console Window, 2 check boxes 2 sliders
Save All Changes to MacroQuest.ini
CheckForPastShowMacroQuestConsole Function to Grab Older MQini and move to new location
Changed Save Location from [Macroquest] to [MQConsole] As now there are 5+ options
Saving: ShowMacroQuestConsole, LockConsoleWindow, NoTitlebar, AlphaFadeOnStartup, AlphaNormalOnStartup
Autoload all options on Startup
Set Default Values for All Options if Button Pressed

Latest Changes:
Removed all Font and Font Size Changes
Code Cleanup
Removed Some Static Variables
SetNextWindowBGAlpha Seems to be Required to be called every frame If i Put a latch on it NoAlpha Console
Found Flag for ImGui::IsWindowHovered(ImGuiHoveredFlags_ChildWindows) Detects mouse in Child Windows

Knightly1 commented:
Closing this one out since the requested changes haven't been addressed.
